src/utils/distributions.cpp

```diff
@@ -18,3 +19,129 @@ double bayesmix::multi_normal_prec_lpdf(const Eigen::VectorXd &datum,
   double exp = (prec_chol * (datum - mean)).squaredNorm();
   return 0.5 * (base - exp);
 }
+
+double bayesmix::gaussian_mixture_dist(
+    Eigen::VectorXd means1, Eigen::VectorXd vars1, Eigen::VectorXd weights1,
+    Eigen::VectorXd means2, Eigen::VectorXd vars2, Eigen::VectorXd weights2) {
+  double mix1 = 0.0;
+#pragma omp parallel for collapse(2) reduction(+ : mix1)
+  for (int i = 0; i < means1.size(); i++) {
+    for (int j = 0; j < means1.size(); j++) {
+      mix1 += weights1(i) * weights1(j) *
+              std::exp(stan::math::normal_lpdf(means1(i), means1(j),
+                                               vars1(i) + vars1(j)));
+    }
+  }
+
+  double mix2 = 0.0;
+#pragma omp parallel for collapse(2) reduction(+ : mix2)
+  for (int i = 0; i < means2.size(); i++) {
+    for (int j = 0; j < means2.size(); j++) {
+      mix2 += weights2(i) * weights2(j) *
+              std::exp(stan::math::normal_lpdf(means2(i), means2(j),
+                                               vars2(i) + vars2(j)));
+    }
+  }
+
+  double inter = 0.0;
+#pragma omp parallel for collapse(2) reduction(+ : inter)
+  for (int i = 0; i < means1.size(); i++) {
+    for (int j = 0; j < means2.size(); j++) {
+      inter += weights1(i) * weights2(j) *
+               std::exp(stan::math::normal_lpdf(means1(i), means2(j),
+                                                vars1(i) + vars2(j)));
+    }
+  }
+
+  return mix1 + mix2 - 2 * inter;
+}
+
+double bayesmix::gaussian_mixture_dist(std::vector<Eigen::VectorXd> means1,
+                                       std::vector<Eigen::MatrixXd> precs1,
+                                       Eigen::VectorXd weights1,
+                                       std::vector<Eigen::VectorXd> means2,
+                                       std::vector<Eigen::MatrixXd> precs2,
+                                       Eigen::VectorXd weights2) {
```

Comment:
Why use std::vectors of Eigen objects instead of just bigger Eigen objects?

Reply:
There would be no problem in storing the means in a matrix (e.g. rowwise) but the precisions would have to be saved in an Eigen::Tensor. Unfortunately the Tensor module is not yet stable.
Moreover having a vector of objects is more lightweight than having complex tensor classes which implement access operations that we don't really use